### DISCUSSION TITLE
IRMP crashes when trying to transmit if using ESP32 board manager v 2.6 or greater

### DISCUSSION THREAD

**Discussion by danthepilotman** [ukw100/IRMP#81] [Q&A] [answered]
Greetings,

While using Arduino IDE V. 2.1.0 I'm seeing an issue where I get a CPU timer_isr_callback fault when trying to transmit IR using an ESP32 Dev Kit 1 board with ESP32 Board Manager version 2.6 or greater. If I revert the board version to 2.5 it works fine. Code snippet below.

Any help in debugging would be appreciated. Might also call for a bug fix in the next library release version.

***********************************
Error Message:
timer_group: timer_isr_callback_add(236): register interrupt service failed

***********************************
Version: 2.1.0
Date: 2023-04-19T15:31:10.185Z
CLI Version: 0.32.2

Copyright © 2023 Arduino SA
***************************************

EPS32 Board Manager version 2.0.5

****************************************
IRMP Library Version 3.6.
****************************************

/* Web_Server_IR_Remote_2.ino
 *  PinDefinitionsAndMore.h
 *
/* NOTE: Requires combination of ESP32 Board Manager Version 2.5 & IRMP Library Version 3.6.2  */

#include <Arduino.h>
#include <WiFi.h>
#include <WebServer.h>
#include "webpageCode.h"

/*
 * Set library modifiers first to set output pin etc.
 */
#include "PinDefinitionsAndMore.h"
//#define IR_OUTPUT_IS_ACTIVE_LOW
#define IRSND_IR_FREQUENCY          38000
#define IRSND_PROTOCOL_NAMES        1 // Enable protocol number mapping to protocol strings - requires some FLASH.
#define IRSND_SUPPORT_KASEIKYO_PROTOCOL        1

/*
 * After setting the definitions we can include the code and compile it.
 */
#include <irsnd.hpp>

IRMP_DATA irsnd_data;
WebServer server(80);

const char* ssid = "Bobo";
const char* wifi_password = "ryrie9219";

//===============================================================
// This prints out Wi-Fi settings
//===============================================================
void printWifiStatus() {

  // print the SSID of the network you're attached to:
  Serial.print("\r\nSSID: ");
  Serial.println(WiFi.SSID());

  // print your WiFi shield's IP address:
  Serial.print("IP Address: ");
  Serial.println(WiFi.localIP());

  // print the received signal strength:
  Serial.print("signal strength (RSSI):");
  Serial.print(WiFi.RSSI());
  Serial.println(" dBm");    
}

//===============================================================
// This routine is executed when changing the operating mode
//===============================================================
void buttonPress() {
 
 irsnd_data.command = (unsigned long) server.arg("CMD_SEL").toInt();
 
  irsnd_send_data(&irsnd_data, true);
  irsnd_data_print(&Serial, &irsnd_data);
 
  server.send(200, "text/plain", "");
 }

void SendWebsite() { 
  server.send(200, "text/html",webpageCode);
}

void setup() {
    Serial.begin(115200);
#if defined(__AVR_ATmega32U4__) || defined(SERIAL_PORT_USBVIRTUAL) || defined(SERIAL_USB) /*stm32duino*/|| defined(USBCON) /*STM32_stm32*/|| defined(SERIALUSB_PID) || defined(ARDUINO_attiny3217)
    delay(4000); // To be able to connect Serial monitor after reset or power up and before first print out. Do not wait for an attached Serial Monitor!
#endif

    // Just to know which program is running on my Arduino
    Serial.println(F("START " __FILE__ " from " __DATE__ "\r\nUsing library version " VERSION_IRMP));

    irsnd_init();
    irmp_irsnd_LEDFeedback(true); // Enable send signal feedback at LED_BUILTIN

    Serial.println(F("Ready to send IR signals at pin " STR(IRSND_OUTPUT_PIN)));

    /*
     * Send NEC
     */
    irsnd_data.protocol = IRMP_KASEIKYO_PROTOCOL;
    irsnd_data.address = 0x2002;
    irsnd_data.command = 0x8320; // The required inverse of the 8 bit command is added by the send routine.
    irsnd_data.flags = 1; // repeat frame 1 time

    // true = wait for frame and trailing space/gap to end. This stores timer state and restores it after sending.
    if (!irsnd_send_data(&irsnd_data, true)) {
        Serial.println(F("Protocol not found")); // name of protocol is printed by irsnd_data_print()
    }
    irsnd_data_print(&Serial, &irsnd_data);

  WiFi.begin(ssid, wifi_password);
  
  while(WiFi.status()!=WL_CONNECTED) {
     delay(500); 
     Serial.print(".");              }
  
  printWifiStatus();

  server.on("/", SendWebsite);
  server.on("/BTN_PRESS", buttonPress);
  server.begin(); //Start server
  Serial.println("HTTP server started");

}

void loop() {    
  server.handleClient();        
}

/*
 *  PinDefinitionsAndMore.h
 *
 *  Contains pin definitions for IRremote examples for various platforms
 *  as well as definitions for feedback LED and tone() and includes
 *
 *  Copyright (C) 2020-2021  Armin Joachimsmeyer
 *  armin.joachimsmeyer@gmail.com
 *
 *  This file is part of IRMP https://github.com/ukw100/IRMP.
 *
 *  IRMP is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/gpl.html>.
 *
 */

/*
 * Pin mapping table for different platforms
 *
 * Platform     IR input    IR output   Tone      Core/Pin schema
 * --------------------------------------------------------------
 * DEFAULT/AVR  2           3           4
 * ATtinyX5     0|PB0       4|PB4       3|PB3
 * ATtiny167    3|PA3       2|PA2       7|PA7     ATTinyCore
 * ATtiny167    9|PA3       8|PA2       5|PA7     Digispark pro
 * ATtiny3217  18|PA1      19|PA2       0|PA4     MegaTinyCore
 * SAMD21       3           4           5
 * ESP8266     14|D5       12|D6        %
 * ESP32       15           4           %
 * BluePill   PA6         PA7         PA3
 * APOLLO3     11          12           5
 * RP2040       3|GPIO15    4|GPIO16    5|GPIO17
 */
//#define IRMP_MEASURE_TIMING // For debugging purposes.
//
#if defined(ESP8266)
#define FEEDBACK_LED_IS_ACTIVE_LOW // The LED on my board is active LOW
#define IRMP_INPUT_PIN      14 // D5
#define IRSND_OUTPUT_PIN    12 // D6 - D4/2 is internal LED
#define IR_TIMING_TEST_PIN  13 // D7

#define tone(...) void()      // tone() inhibits receive timer
#define noTone(a) void()
#define TONE_PIN            42 // Dummy for examples using it

#elif defined(ESP32)
#define IRMP_INPUT_PIN      15  // D15
#define IRSND_OUTPUT_PIN     4  // D4
#include <Arduino.h>
#define TONE_LEDC_CHANNEL        1  // Using channel 1 makes tone() independent of receiving timer -> No need to stop receiving timer.
void tone(uint8_t _pin, unsigned int frequency){
    ledcAttachPin(_pin, TONE_LEDC_CHANNEL);
    ledcWriteTone(TONE_LEDC_CHANNEL, frequency);
}
void tone(uint8_t _pin, unsigned int frequency, unsigned long duration){
    ledcAttachPin(_pin, TONE_LEDC_CHANNEL);
    ledcWriteTone(TONE_LEDC_CHANNEL, frequency);
    delay(duration);
    ledcWriteTone(TONE_LEDC_CHANNEL, 0);
}
void noTone(uint8_t _pin){
    ledcWriteTone(TONE_LEDC_CHANNEL, 0);
}
#define TONE_PIN                27  // D27 25 & 26 are DAC0 and 1

#elif defined(ARDUINO_ARCH_STM32) || defined(ARDUINO_ARCH_STM32F1)
// BluePill in 2 flavors
// Timer 3 of IRMP blocks PA6, PA7, PB0, PB1 for use by Servo or tone()
#define FEEDBACK_LED_IS_ACTIVE_LOW // The LED on the BluePill is active LOW
#define IRMP_INPUT_PIN          PA6
#define IRMP_INPUT_PIN_STRING   "PA6"
#define IRSND_OUTPUT_PIN        PA7
#define IRSND_OUTPUT_PIN_STRING "PA7"
#define TONE_PIN                PA3
#define IR_TIMING_TEST_PIN      PA5

#elif  defined(__AVR_ATtiny25__) || defined(__AVR_ATtiny45__) || defined(__AVR_ATtiny85__)
#include "ATtinySerialOut.hpp" // Available as Arduino library "ATtinySerialOut"
#define IRMP_INPUT_PIN   0
#define IRSND_OUTPUT_PIN 4 // Pin 2 is serial output with ATtinySerialOut. Pin 1 is internal LED and Pin3 is USB+ with pullup on Digispark board.
#define TONE_PIN         3
//#define IR_TIMING_TEST_PIN 3

#elif defined(__AVR_ATtiny87__) || defined(__AVR_ATtiny167__)
#include "ATtinySerialOut.hpp" // Available as Arduino library "ATtinySerialOut"
// For ATtiny167 Pins PB6 and PA3 are usable as interrupt source.
#  if defined(ARDUINO_AVR_DIGISPARKPRO)
#define IRMP_INPUT_PIN   9 // PA3 - on Digispark board labeled as pin 9
//#define IRMP_INPUT_PIN  14 // PB6 / INT0 is connected to USB+ on DigisparkPro boards
#define IRSND_OUTPUT_PIN 8 // PA2 - on Digispark board labeled as pin 8
#define TONE_PIN         5 // PA7
#define IR_TIMING_TEST_PIN 10 // PA4

#  else
#define IRMP_INPUT_PIN   3
#define IRSND_OUTPUT_PIN 2
#define TONE_PIN         7
#  endif

#elif defined(__AVR_ATtiny88__) // MH-ET Tiny88 board
#include "ATtinySerialOut.hpp" // Available as Arduino library "ATtinySerialOut"
// Pin 6 is TX pin 7 is RX
#define IRMP_INPUT_PIN   3 // INT1
#define IRSND_OUTPUT_PIN 4
#define TONE_PIN         9
#define IR_TIMING_TEST_PIN 8

#elif defined(__AVR_ATtiny1616__)  || defined(__AVR_ATtiny3216__) || defined(__AVR_ATtiny3217__)
#define IRMP_INPUT_PIN   18
#define IRSND_OUTPUT_PIN 19
#define TONE_PIN          0

#elif defined(ARDUINO_ARCH_APOLLO3)
#define IRMP_INPUT_PIN   11
#define IRSND_OUTPUT_PIN 12
#define TONE_PIN         5

#elif defined(ARDUINO_ARCH_MBED) && defined(ARDUINO_ARCH_MBED_NANO) // Arduino Nano 33 BLE, Arduino Nano RP2040 Connect
#define IRMP_INPUT_PIN      3   // GPIO15 Start with pin 3 since pin 2|GPIO25 is connected to LED on Pi pico
#define IRSND_OUTPUT_PIN    4   // GPIO16
#define TONE_PIN            5

#elif defined(ARDUINO_ARCH_RP2040) // Pi Pico with arduino-pico core https://github.com/earlephilhower/arduino-pico
#define IRMP_INPUT_PIN      15  // to be compatible with the Arduino Nano RP2040 Connect (pin3)
#define IRSND_OUTPUT_PIN    16
#define TONE_PIN            17

// If you program the Nano RP2040 Connect with this core, then you must redefine LED_BUILTIN
// and use the external reset with 1 kOhm to ground to enter UF2 mode
#undef LED_BUILTIN
#define LED_BUILTIN          6

#elif defined(TEENSYDUINO)
#define IRMP_INPUT_PIN      2
#define IRSND_OUTPUT_PIN    3
#define TONE_PIN            4
#if !defined(ALTERNATIVE_IR_FEEDBACK_LED_PIN)
#define ALTERNATIVE_IR_FEEDBACK_LED_PIN 6 // E.g. used for examples which use LED_BUILDIN for example output.
#endif
#define IR_TIMING_TEST_PIN  7

#elif defined(__AVR__)
#define IRMP_INPUT_PIN      2 // To be compatible with interrupt example, pin 2 is chosen here.
#define IRSND_OUTPUT_PIN    3
#define TONE_PIN            4
#if !defined(ALTERNATIVE_IR_FEEDBACK_LED_PIN)
#define ALTERNATIVE_IR_FEEDBACK_LED_PIN 6 // E.g. used for examples which use LED_BUILDIN for example output.
#endif
#define IR_TIMING_TEST_PIN 7
// You can alternatively specify the input pin with port and bit number if you do not have the Arduino pin number at hand
//#define IRMP_PORT_LETTER D
//#define IRMP_BIT_NUMBER 2

#elif defined(ARDUINO_ARCH_SAMD)
#define IRMP_INPUT_PIN      2
#define IRSND_OUTPUT_PIN    3
#define TONE_PIN            4
#if !defined(ALTERNATIVE_IR_FEEDBACK_LED_PIN)
#define ALTERNATIVE_IR_FEEDBACK_LED_PIN 6 // E.g. used for examples which use LED_BUILDIN for example output.
#endif
#define IR_TIMING_TEST_PIN  7

// On the Zero and others we switch explicitly to SerialUSB
#define Serial SerialUSB

// Definitions for the Chinese SAMD21 M0-Mini clone, which has no led connected to D13/PA17.
// Attention!!! D2 and D4 are switched on these boards!!!
// If you connect the LED, it is on pin 24/PB11. In this case activate the next two lines.
//#undef LED_BUILTIN
//#define LED_BUILTIN 24 // PB11
// As an alternative you can choose pin 25, it is the RX-LED pin (PB03), but active low.In this case activate the next 3 lines.
//#undef LED_BUILTIN
//#define LED_BUILTIN 25 // PB03
//#define FEEDBACK_LED_IS_ACTIVE_LOW // The RX LED on the M0-Mini is active LOW

#else
#warning Board / CPU is not detected using pre-processor symbols -> using default values, which may not fit. Please extend PinDefinitionsAndMore.h.
// Default valued for unidentified boards
#define IRMP_INPUT_PIN      2
#define IRSND_OUTPUT_PIN    3
#define TONE_PIN            4
#if !defined(ALTERNATIVE_IR_FEEDBACK_LED_PIN)
#define ALTERNATIVE_IR_FEEDBACK_LED_PIN 6 // E.g. used for examples which use LED_BUILDIN for example output.
#endif
#define IR_TIMING_TEST_PIN  7
#endif // defined(ESP8266)

#if defined(__AVR_ATmega4809__) // for standard AVR we manage hardware directly in void enablePCIInterrupt()
#define IRMP_USE_ARDUINO_ATTACH_INTERRUPT
#endif

/*
 * Helper macro for getting a macro definition as string
 */
#if !defined(STR_HELPER)
#define STR_HELPER(x) #x
#define STR(x) STR_HELPER(x)
#endif

**Comment by ArminJo** [comment, marked as answer]
I compiled SendAllProtocols and AllProtocols (#define USE_NO_LCD)  with the latest versions and had no errors!

**Reply by danthepilotman** [reply]
I did not have a compile problem, I had a runtime problem.

After compiling, did you run the programs and attempt transmit IR ?
Also, can you confirm the board manager version and IRMP library versions you used.

**Reply by ArminJo** [reply]
> I did not have a compile problem, I had a runtime problem.

Where exactly did you mention this?

Is SendAllProtocols running out of the box?
I always use the latest versions.
Maybe you should test with the version from this repo, to get the same results.

**Reply by danthepilotman** [reply]
In the original post I mentioned that I am seeing a CPU timer_isr_callback fault when attempting to transmit. I'm only interested in KASEIKYO_PROTOCOL 1 and am using version 3.6.2 of the IRMP library.

**Reply by ArminJo** [reply]
Is SendAllProtocols running out of the box?

**Reply by danthepilotman** [reply]
Let me try the SendAllProtocols in my configuration and see what happens.

**Reply by danthepilotman** [reply]
Okay, so I updated my ESP32 Board manager to V 2.9 and compiled the 3.6.2 version of the IRMP SendAllProtocols example. As expected I get the following serial port output:

START C:\Users\sapy4\AppData\Local\Temp\.arduinoIDE-unsaved2023429-17796-5avady.4tprj\SendAllProtocols\SendAllProtocols.ino from May 29 2023
Using library version 3.6.2
Send sample frequency=19000 Hz
Ready to send IR signals at pin 4
Protocol=SIRCS  Address=0x1 Command=0x1
E (3135) timer_group: timer_isr_callback_add(236): register interrupt service failed

**Reply by ArminJo** [reply]
Please change lines 251 ff. in IRTimer.hpp to:    
```
// Use timer with 1 microsecond resolution, main clock is 80MHZ
    if(sReceiveAndSendInterruptTimer == 0) {
        sReceiveAndSendInterruptTimer = timerBegin(3, 80, true);
        timerAttachInterrupt(sReceiveAndSendInterruptTimer, irmp_timer_ISR, false); // false -> level interrupt, true -> edge interrupt, but this is not supported :-(
        timerAlarmWrite(sReceiveAndSendInterruptTimer, ((getApbFrequency() / 80) + (IR_INTERRUPT_FREQUENCY / 2)) / IR_INTERRUPT_FREQUENCY, true);
    }
    timerAlarmEnable(sReceiveAndSendInterruptTimer);
```

**Reply by danthepilotman** [reply]
Modified the IRTimer.hpp file as per you instructions above. Compiled and ran program using IRMP v 3.6.2 and ESP32 Board Manager v 2.9.

Happy to report that all works correctly now !

 Thank you so much for assisting me with this issue. 

So will the fix you recommended be merged in to the next IRMP library release ?

**Reply by ArminJo** [reply]
Thanks for reporting 🥇 
> So will the fix you recommended be merged in to the next IRMP library release ?

Done!
I opened issue #83 for this bug and will delete this discussion in a few days.
